File: 0875-koko-eating-bananas/0875-koko-eating-bananas.py
class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        
        def canEatAll(speed):
            hours = 0
            for pile in piles:
                hours += (pile + speed - 1) // speed  # ceil(pile / speed)
            return hours <= h
        
        left, right = 1, max(piles)
        ans = right
        
        while left <= right:
            mid = (left + right) // 2
            if canEatAll(mid):
                ans = mid
                right = mid - 1
            else:
                left = mid + 1
        
        return ans

File: 0875-koko-eating-bananas/0875-koko-eating-bananas.cpp
class Solution {
public:
    bool canEat(vector<int>& piles, int speed, int h) {
        long long hours = 0;
        for (int pile : piles) {
            hours += (pile + speed - 1) / speed;   // ceil(pile / speed)
            if (hours > h) return false;
        }
        return hours <= h;
    }

    int minEatingSpeed(vector<int>& piles, int h) {
        int left = 1;
        int right = *max_element(piles.begin(), piles.end());
        int ans = right;

        while (left <= right) {
            int mid = left + (right - left) / 2;
            
            if (canEat(piles, mid, h)) {
                ans = mid;         // mid works, try slower speed
                right = mid - 1;
            } else {
                left = mid + 1;    // mid too slow
            }
        }
        return ans;
    }
};
